exporter/signalfx/signalfx.go

```diff
@@ -0,0 +1,293 @@
+			metricName:     sanitize(vd.View.Name),
+			metricType:     "cumulative_counter",
+			metricValueInt: int64(data.Value),
+			timestamp:      time.Now(),
+			dimensions:     buildDimensions(row.Tags),
+		}
+		return metric
+	case *view.LastValueData:
+		metric := signalFxMetric{
+			metricName:  sanitize(vd.View.Name),
+			metricType:  "gauge",
+			metricValue: float64(data.Value),
+			timestamp:   time.Now(),
+			dimensions:  buildDimensions(row.Tags),
+		}
+		return metric
+	default:
+		e.opts.onError(fmt.Errorf("aggregation %T is not yet supported", data))
+		return signalFxMetric{}
+	}
+}
+
+// extractViewData extracts stats data
+func extractViewData(vd *view.Data, e *Exporter) {
+	for _, row := range vd.Rows {
+		metric := e.c.toMetric(vd.View, row, vd, e)
+		go sendRequest(e, metric)
+	}
+}
+
+// buildDimensions uses the tag values and keys to create
+// the dimensions used by SignalFx
+func buildDimensions(t []tag.Tag) map[string]string {
+	values := make(map[string]string)
+	for _, t := range t {
+		values[t.Key.Name()] = t.Value
+	}
+	return values
+}
+
+type collector struct {
+	opts Options
+	mu   sync.Mutex // mu guards all the fields.
+
+	// viewData are accumulated and atomically
+	// appended to on every Export invocation, from
+	// stats. These views are cleared out when
+	// Collect is invoked and the cycle is repeated.
+	viewData map[string]*view.Data
+
+	registeredViewsMu sync.Mutex
+
+	registeredViews map[string]string
+}
+
+// addViewData assigns the view data to the correct view
+func (c *collector) addViewData(vd *view.Data) {
+	c.registerViews(vd.View)
+	sig := viewSignature(vd.View.Name, vd.View)
+
+	c.mu.Lock()
+	c.viewData[sig] = vd
+	c.mu.Unlock()
+}
+
+type signalFxMetric struct {
+	metricType     string
+	metricName     string
+	metricValue    float64
+	metricValueInt int64
+	timestamp      time.Time
+	dimensions     map[string]string
+	buckets        []int64
+}
+
+// newCollector returns a collector struct
+func newCollector(opts Options) *collector {
+	return &collector{
+		opts:            opts,
+		registeredViews: make(map[string]string),
+		viewData:        make(map[string]*view.Data),
+	}
+}
+
+// viewName builds a unique name composed of the namespace
+// and the sanitized view name
+func viewName(namespace string, v *view.View) string {
+	var name string
+	if namespace != "" {
+		name = namespace + "_"
+	}
+	return name + sanitize(v.Name)
+}
+
+// viewSignature builds a signature that will identify a view
+// The signature consists of the namespace, the viewName and the
+// list of tags. Example: Namespace_viewName-tagName...
+func viewSignature(namespace string, v *view.View) string {
+	var buf bytes.Buffer
+	buf.WriteString(viewName(namespace, v))
+	for _, k := range v.TagKeys {
+		buf.WriteString("-" + k.Name())
+	}
+	return buf.String()
+}
+
+// sendRequest sends a package of data containing a metric
+func sendRequest(e *Exporter, data signalFxMetric) {
+	client := sfxclient.NewHTTPSink()
```

Comment:
You should be re-using the same HTTPSink, because it contains an http.Client instance that will be re-used. Otherwise, every time you call `sendRequest()` a new HTTP connection needs to be established with SignalFx.

Reply:
done